src/Control/Monad/Metrics.hs

```diff
@@ -256,10 +260,14 @@ lookupOrCreate
     => (Metrics -> IORef (HashMap k a)) -> (k -> EKG.Store -> IO a) -> k -> m a
 lookupOrCreate getter creator name = do
     ref <- liftM getter getMetrics
-    container <- liftIO $ readIORef ref
-    case HashMap.lookup name container of
-        Nothing -> do
-            c <- liftIO . creator name =<< liftM _metricsStore getMetrics
-            liftIO $ modifyIORef ref (HashMap.insert name c)
-            return c
-        Just c -> return c
+    -- unsafeInterleaveIO is used here to defer creating the metric into
+    -- the 'atomicModifyIORef'' function. We lazily create the value here,
+    -- and the resulting IO action only gets run to create the metric when
+    -- the named metric is not present in the map.
+    newMetric <- liftIO . unsafeInterleaveIO . creator name =<< liftM _metricsStore getMetrics
```

Comment:
@parsonsmatt Any reason you didn't go with `unsafePerformIO` in the end?

Reply:
the travis ci build was going really slowly and I just wanted to get it to pass so I could merge and release